src/openrct2/ride/Vehicle.cpp

```diff
@@ -9304,113 +9304,85 @@ int32_t Vehicle::UpdateTrackMotionMiniGolf(int32_t* outStation)
         }
     }
 
-    regs.eax = 0;
-    regs.dx = 0;
-    regs.ebx = 0;
+    int32_t sumAcceleration = 0;
+    int32_t numVehicles = 0;
     uint16_t totalMass = 0;
 
     for (Vehicle* vehicle = this;;)
     {
-        regs.ebx++;
-        regs.dx |= vehicle->update_flags;
+        numVehicles++;
         totalMass += vehicle->mass;
-        regs.eax += vehicle->acceleration;
-        regs.si = vehicle->next_vehicle_on_train;
-        if (static_cast<uint16_t>(regs.si) == SPRITE_INDEX_NULL)
+        sumAcceleration += vehicle->acceleration;
+        if (vehicle->next_vehicle_on_train == SPRITE_INDEX_NULL)
         {
             break;
         }
-        vehicle = GET_VEHICLE(static_cast<uint16_t>(regs.si));
+        vehicle = GET_VEHICLE(vehicle->next_vehicle_on_train);
     }
 
-    regs.eax /= regs.ebx;
-    regs.ecx = (regs.eax * 21) >> 9;
-    regs.eax = velocity >> 12;
-    regs.ecx -= regs.eax;
-    regs.ebx = velocity;
-    regs.edx = velocity >> 8;
-    regs.edx *= regs.edx;
-    if (regs.ebx < 0)
-    {
-        regs.edx = -regs.edx;
-    }
-    regs.edx >>= 4;
-    regs.eax = regs.edx / totalMass;
-    regs.ecx -= regs.eax;
+    int32_t newAcceleration = (sumAcceleration / numVehicles * 21) >> 9;
+    newAcceleration -= velocity >> 12;
+    newAcceleration -= (((velocity >> 8) * abs(velocity >> 8)) >> 4) / totalMass;
 
-    if (!(vehicleEntry->flags & VEHICLE_ENTRY_FLAG_POWERED))
-    {
-        goto loc_6DD069;
-    }
-    if (vehicleEntry->flags & VEHICLE_ENTRY_FLAG_POWERED_RIDE_UNRESTRICTED_GRAVITY)
+    if (vehicleEntry->flags & VEHICLE_ENTRY_FLAG_POWERED)
     {
-        regs.eax = speed * 0x4000;
-        if (regs.eax < velocity)
+        int32_t poweredAcceleration = speed << 14;
+        if (!(vehicleEntry->flags & VEHICLE_ENTRY_FLAG_POWERED_RIDE_UNRESTRICTED_GRAVITY) || poweredAcceleration >= velocity)
         {
-            goto loc_6DD069;
-        }
-    }
-    regs.eax = speed;
-    regs.bx = track_type >> 2;
-    regs.ebx = regs.eax;
-    regs.eax <<= 14;
-    regs.ebx *= totalMass;
-    regs.ebx >>= 2;
-    if (update_flags & VEHICLE_UPDATE_FLAG_REVERSING_SHUTTLE)
-    {
-        regs.eax = -regs.eax;
-    }
-    regs.eax -= velocity;
-    regs.edx = powered_acceleration;
-    regs.edx <<= 1;
-    regs.eax *= regs.edx;
-    regs.eax = regs.eax / regs.ebx;
-
-    if (!(vehicleEntry->flags & VEHICLE_ENTRY_FLAG_WATER_RIDE))
-    {
-        goto loc_6DD054;
-    }
+            if (update_flags & VEHICLE_UPDATE_FLAG_REVERSING_SHUTTLE)
+            {
+                poweredAcceleration = -poweredAcceleration;
+            }
+            poweredAcceleration -= velocity;
+            poweredAcceleration = poweredAcceleration * (powered_acceleration << 1) / ((speed * totalMass) >> 2);
```

Comment:
Having two variables that only differ by casing is not nice.

Reply:
I took this name from `static void vehicle_update_track_motion_powered_ride_acceleration`, and this problem exists there as well (powered_acceleration is a member variable of Vehicle)

In general this problem exists all over the Vehicle.cpp file ☹️ For example, I see this in many places:
```
uint32_t trackType = track_type >> 2
```
I believe we should keep this name as is, for keeping consistency with the other function, and file an issue to take a look at the more generic problem.

Reply:
Yes, certainly file that issue.

I would even prefer fixing the naming and only _then_ revisiting this, but I'm not going to force you to do that - I don't want to scare away contributors. ;-)